### PR TITLE
Generalise the source list features

### PR DESCRIPTION
**Pull request by ianthehat**
GoEmbed -> GoSources which now holds a list of GoSource entries, each of which
holds most of the data that GoEmbed used to have.
This is in preparation for cleaning up cover and fixing the pure/cgo dichotomy.

### CODE REVIEW

File: go/core.rst [hunk 1]
```diff
@@ -94,7 +94,7 @@ Attributes
 | :param:`embed`             | :type:`label_list`          | :value:`None`                         |
 +----------------------------+-----------------------------+---------------------------------------+
 | List of Go libraries this test library directly.                                                 |
-| These may be go_library rules or compatible rules with the GoEmbed_ provider.                    |
+| These may be go_library rules or compatible rules with the GoSources_ provider.                    |
```
Comment (jayconrod) [inline go/core.rst]: Fix spacing, here and below. These tables aren't rendering anymore.

File: go/private/providers.bzl [hunk 2]
```diff
@@ -21,12 +21,72 @@ GoPackage = provider()
...
+      cgo_deps = cgo_deps,
+      cgo_exports = cgo_exports,
+      cgo_archive = cgo_archive,
+  )
+
+def _source_new(**kwargs):
+  """Creates a new GoSources from a collection of values."""
+  return GoSources(entries = [_source_build_entry(**kwargs)])
+
+def _source_merge(source):
```
Comment (jayconrod) [inline go/private/providers.bzl]: Consider renaming to `sources_list` or something to express this is a collection of `GoSources`. Before I read the doc string, I thought it was a single `GoSource` provider.

File: go/toolchains.rst [hunk 3]
```diff
@@ -490,7 +490,7 @@ It returns GoLibrary_.
 | An optional CgoInfo provider for this library.                                                   |
 | There may be at most one of these among the library and its embeds.                              |
 +--------------------------------+-----------------------------+-----------------------------------+
-| :param:`embed`                 | :type:`GoEmbed iterable`    | :value:`[]`                       |
+| :param:`embed`                 | :type:`GoSources iterable`    | :value:`[]`                       |
```
Comment (jayconrod) [inline go/toolchains.rst]: Table alignment (here and below)

File: go/private/actions/binary.bzl [hunk 4]
```diff
@@ -57,4 +47,4 @@ def emit_binary(ctx, go_toolchain,
       x_defs=x_defs,
   )
 
-  return golib
+  return golib, goembed, goarchive
```
Comment (jayconrod) [inline go/private/actions/binary.bzl]: `goembed` should be renamed to `gosources` here.

File: go/private/actions/library.bzl [hunk 5]
```diff
@@ -31,94 +33,48 @@ load("@io_bazel_rules_go//go/private:actions/archive.bzl",
 def emit_library(ctx, go_toolchain,
     mode = None,
     importpath = "",
-    embed = [],
+    source = None,
```
Comment (jayconrod) [inline go/private/actions/library.bzl]: Consider renaming to `gosources` for consistency with other providers.

File: go/private/rules/binary.bzl [hunk 6]
```diff
@@ -35,21 +36,24 @@ def _go_binary_impl(ctx):
     go_toolchain = ctx.toolchains["@io_bazel_rules_go//go:bootstrap_toolchain"]
 
   executable = ctx.outputs.executable
-  golib = go_toolchain.actions.binary(ctx, go_toolchain,
+  golib, goembed, goarchive = go_toolchain.actions.binary(ctx, go_toolchain,
```
Comment (jayconrod) [inline go/private/rules/binary.bzl]: Rename `goembed` to `gosources`

**Comment by ianthehat** [comment]
Jenkins test this please

**Pull request merged**